pkg/logs/internal/decoder/auto_multiline_detection/incremental_json_validator_test.go

```diff
@@ -0,0 +1,220 @@
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			decoder := NewIncrementalJSONValidator()
+			for i, input := range tt.inputs {
+				assert.Equal(t, tt.expected[i], decoder.Write([]byte(input)))
+			}
+		})
+	}
+}
+
+func TestPrettyPrintedJson(t *testing.T) {
+	jsonString := `{
+    "name": "test",
+    "count": 42,
+    "enabled": true,
+    "tags": [
+        "tag1",
+        "tag2"
+    ]
+}`
+
+	lines := strings.Split(jsonString, "\n")
+	decoder := NewIncrementalJSONValidator()
+	for i, line := range lines[:len(lines)-1] {
+		status := decoder.Write([]byte(line))
+		assert.Equal(t, Incomplete, status, "line %d should be incomplete: %s", i, line)
+	}
+	assert.Equal(t, Complete, decoder.Write([]byte(lines[len(lines)-1])))
+}
+
+func TestPrettyPrintedJsonBrokenFormat(t *testing.T) {
+	jsonString := `{
+    "name": "test",
+    "count": 42,
+    "enabled": true,
+    // suddenly not json`
+
+	lines := strings.Split(jsonString, "\n")
+	decoder := NewIncrementalJSONValidator()
+	for i, line := range lines[:len(lines)-1] {
+		status := decoder.Write([]byte(line))
+		assert.Equal(t, Incomplete, status, "line %d should be incomplete: %s", i, line)
+	}
+	assert.Equal(t, Invalid, decoder.Write([]byte(lines[len(lines)-1])))
+}
+
+func TestEdgeCases(t *testing.T) {
+	tests := []struct {
+		name     string
+		inputs   []string
+		expected []JSONState
+	}{
+		{
+			name:     "empty object",
+			inputs:   []string{"{", "}"},
+			expected: []JSONState{Incomplete, Complete},
+		},
+		{
+			name:     "empty array",
+			inputs:   []string{`{"arr":`, `[`, `]`, `}`},
+			expected: []JSONState{Incomplete, Incomplete, Incomplete, Complete},
+		},
+		{
+			name:     "escaped quotes in strings",
+			inputs:   []string{`{"key":`, `"value \"quoted\" text"`, `}`},
+			expected: []JSONState{Incomplete, Incomplete, Complete},
+		},
+		{
+			name:     "unicode escapes",
+			inputs:   []string{`{"key":`, `"\u0041\u0042\u0043"`, `}`},
+			expected: []JSONState{Incomplete, Incomplete, Complete},
+		},
+		{
+			name:     "whitespace heavy",
+			inputs:   []string{" { ", ` "key" : `, ` "value" `, " } "},
+			expected: []JSONState{Incomplete, Incomplete, Incomplete, Complete},
+		},
+		{
+			name:     "Invalid JSON with array",
+			inputs:   []string{`{`, `[`, `]`, `}`},
+			expected: []JSONState{Incomplete, Invalid, Invalid, Invalid},
+		},
+		{
+			name:     "Incomplete array",
+			inputs:   []string{`{ "arr":`, `[`, `[`, `}`},
+			expected: []JSONState{Incomplete, Incomplete, Incomplete, Invalid},
+		},
+		{
+			name:     "Incomplete nested array",
+			inputs:   []string{`{ "arr":`, `[`, `[`, `]`, `}`},
+			expected: []JSONState{Incomplete, Incomplete, Incomplete, Incomplete, Invalid},
+		},
+		{
+			name:     "Complete nested array",
+			inputs:   []string{`{ "arr":`, `[`, `[`, `]`, `]`, `}`},
+			expected: []JSONState{Incomplete, Incomplete, Incomplete, Incomplete, Incomplete, Complete},
+		},
+		{
```

Comment:
Maybe another test case with a complete JSON object plus extra text might be worth adding?

Reply:
Added an additional test 😄